File: zig/zvk/old/instance.zig
//:__________________________________________________________________
//  zvk  |  Copyright (C) Ivan Mar (sOkam!)  |  GNU GPLv3 or later  :
//:__________________________________________________________________
pub const Instance = @This();
// @deps std
const std = @import("std");
// @deps External
const vk   = @import("../../lib/vulkan.zig");
const glfw = @import("../../lib/glfw.zig");
// @deps zvk
const zvk = struct {
  const Allocator = @import("./allocator.zig").Allocator;
  const Loader    = @import("./loader.zig").Loader;
  const app       = @import("./application.zig").app;
};

//______________________________________
// @section Instance
//____________________________
A    :zvk.Allocator,
ct   :vk.Instance,
cfg  :Instance.Cfg,
pub const Cfg = vk.InstanceCreateInfo;

pub fn create (
    loader  : *zvk.Loader,
    appCfg  : zvk.app.Cfg,
    A       : zvk.Allocator,
  ) !Instance {
  errdefer A.zig.destroy(loader.instance);
  // Get Extensions
  var extCount :u32= 0;
  const exts = glfw.vk.instance.getExts(&extCount);
  // Get Layers
  // Create the result
  var result = Instance{
    .A   = A,
    .ct  = undefined,
    .cfg = Instance.Cfg{
      .flags                      = .{},
      .p_application_info         = &appCfg,
      .enabled_layer_count        = 0,
      .pp_enabled_layer_names     = null,
      .enabled_extension_count    = extCount,
      .pp_enabled_extension_names = @ptrCast(exts),
      }, //:: .cfg
    }; //:: result
  try loader.loadInstance(&result);
  return result;
}
pub fn destroy (I :*Instance, A :std.mem.Allocator) void { A.destroy(I.ct); }

File: nim/nvulkan/instance.nim
#:__________________________________________________________________
#  ndk  |  Copyright (C) Ivan Mar (sOkam!)  |  GNU GPLv3 or later  :
#:__________________________________________________________________
# @deps C Vulkan
{.pragma: vulkan, header:"vulkan/vulkan.h".}
# @deps nvulkan
from ./base as vk import nil
import ./stype
import ./strings
import ./result as R
import ./allocator
import ./application as App

#_____________________________
type Instance * = vk.Instance

#_____________________________
type Flags_C *{.importc:"VkInstanceCreateFlags", vulkan.}= uint32
type Flag *{.importc:"VkInstanceCreateFlagBits", vulkan, pure, size:sizeof(int32).}= enum
  portability
type Flags * = set[Flag]
converter toFlags *(flag :Flag) :Flags= {flag}
converter toFlag  *(flags :Flags_C) :Flags= cast[Flags](flags)

#_____________________________
type Cfg *{.importc:"VkInstanceCreateInfo", vulkan.}= object
  sType       {.importc:"sType"                  .}:StructureType  = stype.InstanceCfg  # VkStructureType  sType;
  next       *{.importc:"pNext"                  .}:pointer        = nil                # const void*      pNext;
  flags      *{.importc:"flags"                  .}:Flags          # VkInstanceCreateFlags       flags;
  appCfg     *{.importc:"pApplicationInfo"       .}:ptr App.Cfg    # const VkApplicationInfo*    pApplicationInfo;
  layerCount *{.importc:"enabledLayerCount"      .}:uint32         # uint32_t                    enabledLayerCount;
  layers     *{.importc:"ppEnabledLayerNames"    .}:StringList     # const char* const*          ppEnabledLayerNames;
  extCount   *{.importc:"enabledExtensionCount"  .}:uint32         # uint32_t                    enabledExtensionCount;
  exts       *{.importc:"ppEnabledExtensionNames".}:StringList     # const char* const*          ppEnabledExtensionNames;

proc create *(
    cfg    : ptr instance.Cfg;
    A      : ptr Allocator;
    result : ptr vk.Instance;
  ) :Result {.importc:"vkCreateInstance", vulkan, cdecl.}

proc destroy *(
    I : vk.Instance;
    A : ptr Allocator;
  ) :void {.importc:"vkDestroyInstance", vulkan, cdecl.}

